src/tactic/move_add.lean

```diff
@@ -0,0 +1,224 @@
+The only pre-processing that we do to the lists is that we ignore terms of either one of the two
+lists in `ll` that do not appear in `get_summands e`.
+
+We use this function for expressions in an additive commutative semigroup. -/
+meta def sorted_sum_with_rel
+  (hyp : option name) (rel : expr → expr → bool) (ll : list expr × list expr) (e : expr) :
+  tactic unit :=
+match (get_summands e).sort_with_ends ll rel with
+| eₕ::es := do
+  e' ← es.mfoldl (λ eₗ eᵣ, mk_app `has_add.add [eₗ, eᵣ]) eₕ,
+  e_eq ← mk_app `eq [e, e'],
+  n ← get_unused_name,
+  assert n e_eq,
+  e_eq_fmt ← pp e_eq,
+  reflexivity <|>
+    `[{ simp only [add_comm, add_assoc, add_left_comm], done, }] <|>
+    -- `[{ abel, done, }] <|> -- this works too. it's more robust but also a bit slower
+      fail format!"failed to prove:\n\n{e_eq_fmt}",
+  h ← get_local n,
+  match hyp with
+  | some loc := do
+    get_local loc >>= rewrite_hyp h,
+    tactic.clear h
+  | none     := do
+    rewrite_target h,
+    tactic.clear h
+  end
+| [] := skip
+end
+
+/-- Partially traverses an expression in search for a sum of terms.
+When `recurse_on_expr` finds a sum, it sorts it using `sorted_sum_with_rel`. -/
+meta def recurse_on_expr
+  (ll : list expr × list expr) (hyp : option name) (rel : expr → expr → bool) : expr → tactic unit
+| e@`(%%_ + %%_)     := sorted_sum_with_rel hyp rel ll e
+| (expr.lam _ _ _ e) := recurse_on_expr e
+| (expr.pi  _ _ _ e) := recurse_on_expr e
+| e                  := e.get_app_args.mmap' recurse_on_expr
+
+/-- Calls `recurse_on_expr` with the right expression, depending on the tactic location. -/
+meta def move_add_aux (rel : expr → expr → bool) (ll : list expr × list expr) :
+  option name → tactic unit
+| (some hyp) := get_local hyp >>= infer_type >>= recurse_on_expr ll hyp rel
+| none       := target >>= recurse_on_expr ll none rel
+
+end with_rel
+
+/--  A version of `move_add_aux` that allows failure, if `allow_failure = tt`. -/
+meta def move_add_core (allow_failure : bool) (rel : expr → expr → bool)
+  (ll : list expr × list expr) (hyp : option name) :
+  tactic unit :=
+if allow_failure then move_add_aux rel ll hyp <|> skip
+else move_add_aux rel ll hyp
+
+/-- `move_add_arg` is a single elementary argument that `move_add` takes for the
+variables to be moved.  It is either a `pexpr`, or a `pexpr` preceded by a `←`. -/
+meta def move_add_arg (prec : nat) : lean.parser (bool × pexpr) :=
+prod.mk <$> (option.is_some <$> (tk "<-")?) <*> parser.pexpr prec
+
+/-- `move_pexpr_list_or_texpr` is either a list of `move_add_arg`, possibly empty, or a single
+`move_add_arg`. -/
+meta def move_pexpr_list_or_texpr : lean.parser (list (bool × pexpr)) :=
+list_of (move_add_arg 0) <|> list.ret <$> (move_add_arg tac_rbp) <|> return []
+
+/--
+Let `rel : expr → expr → bool` be a relation on `expr`s.
+Calling `move_add_with_rel rel [a, ← b, c, ← d]`, recursively looks inside the goal for
```

Comment:
It would be nice to factor out a non-recursive version, which would be handy in conv mode

Reply:
Eric, I like this idea very much: I'll look into it!

I am still trying to get the tactic to behave as I want and I think that I am close now!

Reply:
Eric, this means that the tactic should work directly on an expression, assuming that the expression begins being `has_add.add _ _`?